Add pre-commit to improve commit quality

rehsack opened the pull request:
See https://pre-commit.com/ for details

### All Submissions:

* [x] Did you update the `changelog.adoc`?
       Not yet, it's a draft and needs discussion
* [x] Does your PR affect the documentation?
       Maybe the documentation for contributors
* [x] If yes, did you update the documentation or create an issue for updating it?
       Will do when necessary

The source of the documentation can be found in `/src/docs/`.

If you didn't find the time to update docs, please create an issue as reminder to do so.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Your first submission

* [x] Welcome to the list of contributors! If you have any questions, feel free to ask them by creating a new issue
* [x] Have you added your name to the list of [30_community.adoc](https://github.com/docToolchain/docToolchain/blob/ng/src/docs/10_about/30_community.adoc)?


inspiration: https://github.com/stevemao/github-issue-templates

Commit:
feat: add pre-commit to improve commit quality

Add pre-commit with a basic configuration to improve future commits
as well as let uns drawing a baseline for formatting to avoid related
discussions in contributions.

Signed-off-by: Jens Rehsack <jens.rehsack.ext@vaillant-group.com>

rdmueller commented:
Wow. Will the current code pass the pre-commit hooks?

rdmueller reviewed (approved):
Quite a lot of hooks but I am willing to give it a try

rehsack commented:
> Wow. Will the current code pass the pre-commit hooks?

No, but I didn't want start committing the auto-fixes nor fix manually and add a CI workflow when the idea is not welcome.

The most hooks are from [Pre Commit > Supported hooks](https://pre-commit.com/hooks.html) and there from https://github.com/pre-commit/pre-commit-hooks (see README.md).

I would also dig a bit for asciidoctor or Ruby/jRuby/Gradle hooks. But let's start "small".

rdmueller commented:
This is still a Draft. What will be the next steps?

rehsack commented:
I have to find a few minutes to rebase and add some reasonable plugins. Starting with calender week 51 there will be spare tme for doing that. Thanks for reminding me.

rehsack committed:
chore: apply pre-commit rules

Run `pre-commit run -a` and fix found issues (beside bats tests).

Signed-off-by: Jens Rehsack <sno@netbsd.org>

rehsack committed:
chore: ignore Vi & Emacs temp files for git

Ignore temporary files created while working with Emacs or Vi for
working with git on this repository.

Signed-off-by: Jens Rehsack <sno@netbsd.org>

rehsack committed:
docs: add a note regarding to pre-commit

Avoid adding a dedicated workflow since https://pre-commit.ci is likely
satisfying the need of dedicated workflows to prove commit quality.

Signed-off-by: Jens Rehsack <sno@netbsd.org>

rehsack commented:
@rdmueller - beside the issues in the bats tests, I fixed the found issues unless they're fixed automatically. I would prefer using https://pre-commit.ci in favour of an own workflow. I expect this works better.

Maybe https://github.com/DanySK/gradle-pre-commit-git-hooks is worth a look - but it seems to be something completely different approach. Further: I didn't find Gradle/Groovy checkers similar[^1,^2] to Ruby, C++, Shell etc.

I think, having something which proves the shell sources, YAML and JSON files (and schemata) can improve enough for first step. 

^1: Please have a look how https://github.com/nvuillam/npm-groovy-lint looks to you. The patch for `.pre-commit-config.yaml` should be trivial, if you like it.
^2: Check https://github.com/rehsack/docToolchain/tree/feature/add-groovy-to-pre-commit-config how it behaves out-of-the-box

rdmueller committed:
Merge branch 'ng' into feature/add-pre-commit-support

rdmueller commented:
wow. This is a HUGE PR and the build fails. How can we proceed with this?

rehsack commented:
As said - I am not a Ruby nor a Groovy developer, I am just an interested user.
Since no source has been changed, someone with knowledge should look into that ...or guide me - I am willing to learn for many reasons.
I can also ... when it's not again left on side for 5 month, rebase the PR - this isn't to much work (but useless, when wait for next conflicts).

rehsack commented:
I would nowadays looking to lefthook or Nix Flakes instead.
